api/v1beta1/types_common.go

```diff
@@ -0,0 +1,142 @@
+	// Some values for the cloud provider config are inferred from other parts of cluster api provider azure spec, and may not be available for overrides.
+	// See: https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs
+	// Note: All cloud provider config values can be customized by creating the secret beforehand. CloudProviderConfigOverrides is only used when the secret is managed by the Azure Provider.
+	// +optional
+	CloudProviderConfigOverrides *CloudProviderConfigOverrides `json:"cloudProviderConfigOverrides,omitempty"`
+}
+
+// NetworkClass defines the NetworkSpec properties that may be shared across several Azure clusters.
+type NetworkClass struct {
+	// PrivateDNSZoneName defines the zone name for the Azure Private DNS.
+	// +optional
+	PrivateDNSZoneName string `json:"privateDNSZoneName,omitempty"`
+}
+
+// VnetClass defines the VnetSpec properties that may be shared across several Azure clusters.
+type VnetClass struct {
+	// CIDRBlocks defines the virtual network's address space, specified as one or more address prefixes in CIDR notation.
+	// +optional
+	CIDRBlocks []string `json:"cidrBlocks,omitempty"`
+
+	// Tags is a collection of tags describing the resource.
+	// +optional
+	Tags Tags `json:"tags,omitempty"`
+}
+
+// SubnetClass defines the SubnetSpec properties that may be shared across several Azure clusters.
+type SubnetClass struct {
+	// Role defines the subnet role (eg. Node, ControlPlane)
+	// +kubebuilder:validation:Enum=node;control-plane;bastion
+	Role SubnetRole `json:"role"`
+
+	// CIDRBlocks defines the subnet's address space, specified as one or more address prefixes in CIDR notation.
+	// +optional
+	CIDRBlocks []string `json:"cidrBlocks,omitempty"`
+}
+
+// LoadBalancerClass defines the LoadBalancerSpec properties that may be shared across several Azure clusters.
+type LoadBalancerClass struct {
+	// +optional
+	SKU SKU `json:"sku,omitempty"`
+	// +optional
+	FrontendIPs []FrontendIP `json:"frontendIPs,omitempty"`
+	// +optional
+	Type LBType `json:"type,omitempty"`
+	// FrontendIPsCount specifies the number of frontend IP addresses for the load balancer.
+	// +optional
+	FrontendIPsCount *int32 `json:"frontendIPsCount,omitempty"`
+	// IdleTimeoutInMinutes specifies the timeout for the TCP idle connection.
+	// +optional
+	IdleTimeoutInMinutes *int32 `json:"idleTimeoutInMinutes,omitempty"`
+}
+
+// SecurityGroupClass defines the SecurityGroup properties that may be shared across several Azure clusters.
+type SecurityGroupClass struct {
+	// +optional
+	SecurityRules SecurityRules `json:"securityRules,omitempty"`
+	// +optional
+	Tags Tags `json:"tags,omitempty"`
+}
+
+// FrontendIPClass defines the FrontendIP properties that may be shared across several Azure clusters.
+type FrontendIPClass struct {
+	// +optional
+	PrivateIPAddress string `json:"privateIP,omitempty"`
+}
+
+// setDefaults sets default values for AzureClusterClass.
+func (acc *AzureClusterClass) setDefaults() {
```

Comment:
What was the reasoning for putting these funcs in types_common.go and not azurecluster_defaults.go?

Reply:
When we add `AzureClusterTemplate` type, I'm guessing we'll be creating a separate `azureclustertemplate_defaults.go` where we'll need to set these defaults again. So, I thought this will be a good place to put and reuse it in both azurecluster_defaults and azureclustertemplate_defaults. This is somewhat an extension of the idea mentioned in [cascading type validations](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1338)